### DISCUSSION TITLE
dotenv problem

### DISCUSSION THREAD

**Discussion by keeskraak** [quasarframework/quasar#9967] [General - Components / Directives / etc]
Hi,

I am fairly new to Quasar, but I am running into a strange problem concerning the dotenv package.

I use quasar 2.0 and created a brand new project by running "quasar create test-project"
After creation I run "quasar dev" and everything is running fine.
Then I add the dotenv package by running "npm install dotenv" and add it to quasar.conf.js:

```
    build: {
      env: require('dotenv').config().parsed
    }
```
Next I created a .env file in the root-directory:
```
TEST=1
```

When I use this variable in my code, it works fine, like:

```
computed: {
  isTestSet() {
      return process.env.TEST === '1'
  }
}
```

But, when I use a variable which is NOT defined in the .env file, like:
```
computed: {
  isTestSet() {
      return process.env.TEST2 === '1'
  }
}
```

In this case I get a weird error: 
```
Uncaught (in promise) ReferenceError: process is not defined
```

So, it only works if the referenced variable is defined in the .env file.
Otherwise I get this error saying "Uncaught (in promise) ReferenceError: process is not defined"

As far as I know this did work in previous versions?

What do I do wrong?

Thank you very much.

**Comment by justinbeaty** [comment]
@keeskraak I'm not sure how long ago or what previous version you had used, but the behavior changed for security reasons about 7-8 months ago. See #8084 and #8361

**Reply by hawkeye64** [reply]
That is correct. You can't `console.log(process.env)` because Quasar does some special manipulation to obfuscate your data from prying eyes. The whole object path **must** exist when using it.

**Reply by keeskraak** [reply]
Ok, thank you very much.

I normally check if an environment-variable is set. If not I use a default value.
So, this is no longer possible? Or is there another way around this?

For example
```
getDelay() {
   return process.env.DELAY ? process.env.DELAY : 5000
}
```

If DELAY is not defined as environment variable, this would result in "ReferenceError: process is not defined".
Is there any alternative solution?

**Reply by justinbeaty** [reply]
@keeskraak You can instead set default values in quasar.conf.js, as it has access to all env vars. Something like this:

```
module.exports = function (ctx) {
  return {
    // ...
    build: {
      // ...
      env: {
        DELAY: process.env.DELAY ? process.env.DELAY : 5000,
        SOMETHING_ELSE: 'foobar',
        // etc
      },
    },
  }
}
```

Then, `process.env.DELAY` is always defined and you also get all default values in one central place.

**Edit:** Since I forgot you used dotenv, you can modify it like so:

```
env: {
   DELAY: 5000, // the default value
   SOMETHING_ELSE: 'foobar',
   ...(require('dotenv').config().parsed),
},
```

That way you define the defaults, but overwrite them with the spread operator. Or use something like: https://www.npmjs.com/package/dotenv-defaults

**Reply by keeskraak** [reply]
Wauw, thank you so much. This works like a charm!
